openssl/openssl.go

```diff
@@ -168,44 +170,43 @@ func FIPS() bool {
 // SetFIPS enables or disables FIPS mode.
 //
 // It implements the following provider fallback logic for OpenSSL 3:
-//    - The "fips" provider is loaded if enabled=true and no loaded provider matches "fips=yes".
-//    - The "default" provider is loaded if enabled=false and no loaded provider matches "fips=no".
+//   - The "fips" provider is loaded if enabled=true and no loaded provider matches "fips=yes".
+//   - The "default" provider is loaded if enabled=false and no loaded provider matches "fips=no".
+//
 // This logic allows advanced users to define their own providers that match "fips=yes" and "fips=no" using the OpenSSL config file.
 func SetFIPS(enabled bool) error {
+	var mode C.int
+	if enabled {
+		mode = C.int(1)
+	} else {
+		mode = C.int(0)
+	}
 	switch vMajor {
 	case 1:
-		var mode C.int
-		if enabled {
-			mode = C.int(1)
-		} else {
-			mode = C.int(0)
-		}
 		if C.go_openssl_FIPS_mode_set(mode) != 1 {
 			return newOpenSSLError("openssl: FIPS_mode_set")
 		}
 		return nil
 	case 3:
-		var props, provName *C.char
+		var provName *C.char
 		if enabled {
-			props = propFipsYes
 			provName = providerNameFips
 		} else {
-			props = propFipsNo
 			provName = providerNameDefault
 		}
 		// Check if there is any provider that matches props.
-		if !providerAvailable(props) {
+		if C.go_openssl_OSSL_PROVIDER_available(nil, provName) == 0 {
 			// If not, fallback to provName provider.
```

Comment:
This doesn't seem to line up with the behavior mentioned in the `SetFIPS` func comment--it seems like there isn't room for a config-file-based provider to sneak in anymore.

Reply:
Did that on purpose, but forgot to remove the comment. I now think that supporting FIPS providers not named `fips` (as the official one) is overthinking, and the implications of that are unknown to me, I haven't seen that support elsewhere. For example, [nodejs](https://github.com/nodejs/node/blob/607c8f4eacd5dded97a33f0da49da0aed7404187/src/crypto/crypto_util.cc#L120-L138) checks if the `fips` provider can be loaded and if is enabled by default, and [ruby](https://github.com/ruby/ruby/blob/d5ef373b1194bac64784ae316d125d7a2cf1988a/ext/openssl/ossl.c#L447-L459) only checks if it is enabled by default.

Reply:
Notice that I've implemented things a bit different in https://github.com/golang-fips/openssl-fips/pull/66. That's because I don't want to break the go-crypto-openssl API, in case we decide to backport this.